transport/testing/reqres.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/yarpc/yarpc-go/transport"
+)
+
+// RequestMatcher may be used in gomock argument lists to assert that two
+// requests match.
+//
+// Requests are considered to be matching if:
+//
+// - All their primitive parameters match.
+// - The headers of the received request include all the headers from the
+//   source request. It may include extra headers.
+// - The contents of the request bodies are the same.
+type RequestMatcher struct {
+	t    *testing.T
+	req  *transport.Request
+	body []byte
+}
+
+// NewRequestMatcher constructs a new RequestMatcher from the given testing.T
+// and request.
+//
+// The request's contents are read in their entirety and replaced with a
+// bytes.Reader.
+func NewRequestMatcher(t *testing.T, r *transport.Request) RequestMatcher {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		t.Fatalf("failed to read request body: %v", err)
+		panic(err.Error())
+	}
+
+	// restore a copy of the body so that the caller can still use the request
+	// object
+	r.Body = bytes.NewReader(body)
+	return RequestMatcher{t: t, req: r, body: body}
+}
+
+// TODO: Headers like User-Agent, Content-Length, etc. make their way to the
+// user-level Request object. For now, we're doing the super set check but we
+// should do something more specific once yarpc/yarpc.io#2 is resolved.
+
+// Matches checks if the given object matches the Request provided in
+// NewRequestMatcher.
+func (m RequestMatcher) Matches(got interface{}) bool {
+	l := m.req
+	r, ok := got.(*transport.Request)
+	if !ok {
+		panic(fmt.Sprintf("expected *transport.Request, got %v", got))
+	}
+
+	if l.Caller != r.Caller {
+		m.t.Logf("Caller mismatch: %s != %s", l.Caller, r.Caller)
+		return false
+	}
+
+	if l.Service != r.Service {
+		m.t.Logf("Service mismatch: %s != %s", l.Service, r.Service)
+		return false
+	}
+
+	if l.Procedure != r.Procedure {
+		m.t.Logf("Procedure mismatch: %s != %s", l.Procedure, r.Procedure)
+		return false
+	}
+
+	if l.TTL != r.TTL {
+		m.t.Logf("TTL mismatch: %v != %v", l.TTL, r.TTL)
+		return false
+	}
+
+	if !isSuperSet(l.Headers, r.Headers) {
+		m.t.Logf("Headers mismatch: %v != %v", l.Headers, r.Headers)
+		return false
+	}
+
+	rbody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		m.t.Fatalf("failed to read body: %v", err)
+		return false
+	}
+	r.Body = bytes.NewReader(rbody) // in case it is reused
+
+	if !bytes.Equal(m.body, rbody) {
+		m.t.Logf("Body mismatch: %v != %v", m.body, rbody)
+		return false
+	}
+
+	return true
+}
+
+func (m RequestMatcher) String() string {
+	return fmt.Sprintf("matches request %v with body %v", m.req, m.body)
+}
+
+// isSuperSet checks if the items in l are all also present in r.
+func isSuperSet(l, r map[string]string) bool {
+	// TODO Figure out if headers are case sensitive?
+	rLower := make(map[string]string, len(r))
+	for k, v := range r {
+		rLower[strings.ToLower(k)] = v
+	}
+
+	for k, vl := range l {
+		vr, ok := rLower[strings.ToLower(k)]
+		if !ok || vr != vl {
+			return false
```

Comment:
optional: let's make this method return an `error` and give information here about why it failed (e.g. this header was not found but was expected to be <blah>, or this header did not match).